plugins/workload-management/src/internalClusterTest/java/org/opensearch/plugin/wlm/WlmAutoTaggingIT.java

```diff
@@ -0,0 +1,866 @@
+
+        FeatureType featureType = AutoTaggingRegistry.getFeatureType(WorkloadGroupFeatureType.NAME);
+        createRule(ruleId, "tagging flow test", "logs-tagged*", featureType, workloadGroupId);
+
+        indexDocument(indexName);
+
+        assertBusy(() -> {
+            int completionsBefore = getCompletions(workloadGroupId);
+            client().prepareSearch(indexName).setQuery(QueryBuilders.matchAllQuery()).get();
+            client().admin().indices().prepareRefresh(indexName).get();
+            int completionsAfter = getCompletions(workloadGroupId);
+            assertTrue("Expected completions to increase", completionsAfter > completionsBefore);
+        });
+
+        clearWlmModeSetting();
+    }
+
+    public void testMultipleRulesDoNotInterfereWithEachOther() throws Exception {
+        String index1 = "test_1";
+        String index2 = "test_2";
+        String testIndex = "test_*";
+        String workloadGroupId1 = "wlm_auto_tag_group_1";
+        String workloadGroupId2 = "wlm_auto_tag_group_2";
+        String ruleId1 = "wlm_auto_tag_rule_1";
+        String ruleId2 = "wlm_auto_tag_rule_2";
+
+        setWlmMode("enabled");
+
+        updateWorkloadGroupInClusterState(PUT, createWorkloadGroup("group_1", workloadGroupId1));
+        updateWorkloadGroupInClusterState(PUT, createWorkloadGroup("group_2", workloadGroupId2));
+
+        FeatureType featureType = AutoTaggingRegistry.getFeatureType(WorkloadGroupFeatureType.NAME);
+
+        createRule(ruleId1, "rule for test_1", index1, featureType, workloadGroupId1);
+        createRule(ruleId2, "rule for test_2", index2, featureType, workloadGroupId2);
+
+        indexDocument(index1);
+        indexDocument(index2);
+
+        assertBusy(() -> {
+            int pre1 = getCompletions(workloadGroupId1);
+            int pre2 = getCompletions(workloadGroupId2);
+
+            client().prepareSearch(testIndex).setQuery(QueryBuilders.matchAllQuery()).get();
+
+            int post1 = getCompletions(workloadGroupId1);
+            int post2 = getCompletions(workloadGroupId2);
+
+            assertTrue("Expected completions for group 1 not to increase", post1 == pre1);
+            assertTrue("Expected completions for group 2 not to increase", post2 == pre2);
+        });
+
+        clearWlmModeSetting();
+    }
+
+    public void testTaggingTriggeredAfterRuleUpdate() throws Exception {
+        String workloadGroupId = "wlm_auto_tag_update";
+        String ruleId = "wlm_auto_tag_update_rule";
+        String indexName = "update_index";
+
+        setWlmMode("enabled");
+
+        updateWorkloadGroupInClusterState(PUT, createWorkloadGroup("update_test_group", workloadGroupId));
+
+        FeatureType featureType = AutoTaggingRegistry.getFeatureType(WorkloadGroupFeatureType.NAME);
+        createRule(ruleId, "initial non-matching rule", "random", featureType, workloadGroupId);
+
+        // Wait for rule to be available
+        Thread.sleep(5500);
```

Comment:
Can we please try to avoid using Thread.sleep at all costs? Wait for a condition to be met and proceed otherwise fail the test after some timeout if the condition is not met.